File: tests/test_get_major_eigenvalue/test_get_major_eigenvalue.py

import sys
import os

# 将drr_tree.py文件所在目录添加到系统路径
module_path = '/mnt/ssd/geneplus/develop/liushen/projects/DDRTreepy/DDRTree'  # ddr_tree.py所在的目录
if module_path not in sys.path:
    sys.path.append(module_path)

# 导入ddr_tree模块
from ddr_tree import * # 现在可以导入ddr_tree.py文件

import pandas as pd

    

# 指定CSV文件路径
csv_file_path = '/mnt/ssd/geneplus/develop/liushen/projects/DDRTreepy/tests/data/random_matrix.csv'

# 使用pandas读取CSV文件
df = pd.read_csv(csv_file_path, header=None)  # header=None表示不将第一行作为列名

# 将DataFrame转换为NumPy数组（二维矩阵）
matrix = df.values  # 或使用 df.to_numpy()

# 打印导入的矩阵
# print(matrix)

C = matrix

L = 5  # 设置要检索的前 5 个特征值
top_eigenvalue = get_major_eigenvalue_python(C, L)  # 调用函数获取特征值
print(top_eigenvalue)  # 输出最大特征值

File: tests/test_get_major_eigenvalue/test_get_major_eigenvalue.R
library(irlba)

#' Get the top L eigenvalues
#' @param C data matrix used for eigendecomposition
#' @param L number for the top eigenvalues
#' @import irlba irlba
#' @export
get_major_eigenvalue <- function(C, L) {
    if (L >= min(dim(C))){
        return (base::norm(C, '2')^2);
    }else{
        #message("using irlba")
        initial_v <- as.matrix(qnorm(1:(ncol(C) + 1)/(ncol(C) + 1))[1:ncol(C)])
        eigen_res <- irlba(C, nv = L, v = initial_v)
        return (max(abs(eigen_res$v)))
    }
    #     eig_sort <- sort(V, decreasing = T, index.return = T)
    #     eig_idx <- eig_sort$ix
    #
    #     W <- U[, eig_idx[1:L]]
}


# 指定CSV文件路径
csv_file_path = '/mnt/ssd/geneplus/develop/liushen/projects/DDRTreepy/tests/data/random_matrix.csv'

data <- read.csv(csv_file_path, header = TRUE)  # header = TRUE 表示文件包含列名，可根据实际情况修改

# 将数据框转换为矩阵
matrix_data <- as.matrix(data)

C = matrix_data



L = 5  # 设置要检索的前 5 个特征值
top_eigenvalue = get_major_eigenvalue(C, L)  # 调用函数获取特征值
print(top_eigenvalue)  # 输出最大特征值
